### DISCUSSION TITLE
Conditional Request Forwarding

### DISCUSSION THREAD

**Discussion by noh626** [dotnet/yarp#2295] [Q&A] [answered]
Hi,

I’m currently exploring ways to selectively forward requests either to local controllers or downstream services based on a custom flag. While utilizing the Order property seems to be the most straightforward approach for routes with the same pattern, it only works effectively when the route configuration in YARP and the route for the local controller are identically specific, as the most specific match wins.

Here’s the scenario I’m dealing with:

- I have a local endpoint defined for "controller/action1".

- YARP is configured to proxy "controller/{**remainder}" to a cluster.

With this setup, a request matching "controller/action1" naturally hits the local endpoint. However, I want to have the flexibility to selectively proxy this request based on a custom flag, even when it matches the local route.

Is there a way I can manually control whether a request gets proxied or not within a middleware component? Alternatively, if there’s another approach or strategy that could achieve this selective forwarding, I’m open to suggestions and would appreciate any insights.

**Comment by Tratcher** [comment, marked as answer]
Yes, in the middleware you can check the HttpContext to see if you want to forward. If so, you call IHttpForwarder, else call the next middleware.

This will show you how to get and call the IHttpForwarder.
https://microsoft.github.io/reverse-proxy/articles/direct-forwarding.html

**Reply by noh626** [reply]
Thank you for your response.

Currently, I'm utilizing MapReverseProxy along with a custom implementation of the IForwarderHttpClientFactory interface. This means I'm mostly relying on the default processing pipeline provided by MapReverseProxy.

Based on your suggestion, are you implying that I should create my own custom middleware to replace the default ForwarderMiddleware? And in this custom middleware, I should inspect the HttpContext to determine whether to call IHttpForwarder for forwarding or invoke the next middleware in the pipeline for local handling?

Could you please elaborate a bit more on integrating this into the existing pipeline set up by MapReverseProxy?

**Reply by Tratcher** [reply]
Do not use MapReverseProxy in this case, that requires you to have defined routes. With your own middleware in the main pipeline you let the local routs match and then decide if you want to override them by calling IHttpFowarder.

What were you using from the proxy pipeline? Config load, health checks, load balancing, etc.? How many destinations are you forwarding to?

**Reply by noh626** [reply]
My primary use of the proxy is its core forwarding functionality for now. However, I may have to support additional features such as load balancing and health checks in the future. Currently, I am forwarding traffic to a couple of downstream services.

What I wanted is to have the flexibility to selectively forward traffic either to these remote services or to local controllers based on specific criteria. If I understand correctly, your suggestion is that I should implement my own middleware that performs functions similar to those of the default ForwarderMiddleware, but with additional logic to selectively decide whether to call the next middleware or SendAsync of IHttpForwarder. Is that correct?

**Reply by Tratcher** [reply]
Yes, that's the idea. You shouldn't need any of the content from ForwarderMiddleware, that's all about adapting between the higher level config model and the IHttpFowarder. If you have your own criteria and a list of destinations, you don't need YARPs config model.

See how that works and we can revisit the other features in the future.

**Reply by noh626** [reply]
I understand the approach you've outlined. However, I'd like to know if there's a way to still take advantage of YARP's routing and transformation configuration within my custom middleware. My intention is to potentially use the metadata property in YARP's configuration to inform the logic of my middleware. This way, I could manage selective forwarding while also making use of the sophisticated routing and request transformation setup that YARP provides. How might I go about accessing and utilizing YARP's configured metadata for this purpose?

**Reply by noh626** [reply]
I've made an interesting observation regarding the `Order` field in YARP's configuration that seems to contradict my understanding.

Initially, I believed that the `Order` field would only influence the routing when that path specified in YARP's `RouteConfig` had the same level of specificity as a local route. However, during my testing with the following setup:

- YARP's `RouteConfig`:
`"Route1": {
  "Order": -1,
  "Match": {
    "Path": "abc/123/{**remainder}"
  }
}`
- Local action route: `abc/123/action/{id}`

I expected that a request matching `abc/123/action/{id}` would always be handled by the local action, regardless of the `Order` value in `RouteConfig`. Contrary to this, `Route1` with `Order` set to -1 took precedence over the local route.

The reason I wanted to use the `Metadata` property in YARP's configuration was under the assumption that the `Order` property did not affect routes with varying levels of path specification.

Did I misunderstand the role of `Order` in YARP's routing?

All I wanted are to leverage YARP's cofig model and selectively route traffic to either local controllers or remote APIs. I didn't want to enumerate every single endpoint in YARP's config using the `Order` property. I assumed that `Order` wouldn't apply across routes of different path specificity which led me to consider the `Metadata` property as a potential solution.

**Reply by Tratcher** [reply]
Right, Order is actually checked first, then the path. Note this is not YARP specific, YARP is just plugging into AspNetCore's routing.
https://learn.microsoft.com/en-us/aspnet/core/fundamentals/routing?view=aspnetcore-7.0#url-matching

That's also why it's hard to do dynamically choose between routes, that's not a feature supported by AspNetCore. There are some advanced route constraint extensions, but they're not easy to use. That's why I suggested middleware instead.

**Reply by noh626** [reply]
I thought ASP.NET Core routing mechanism doesn't check the `Order` before the route template specificity and the default behavior is for the routing to match the most specific route template first.

However, it appears to function in the reverse order, with `Order` taking precedence.

This actually simplifies my scenario. By dynamically adjusting the `Order` in YARP's config as needed, I can manage the routing to prefer local endpoints or proxy forwarding accordingly. For example, if a specific endpoint requires a unique forward flag, I could introduce a dedicated route config with an `Order` that takes precedence over a more generic route configuration and YARP's config can be updated during the runtime.

Do you think this is a viable solution?

**Reply by Tratcher** [reply]
If that's dynamic enough for you, great.
